### PR TITLE
Parse Lambda AppSec request bodies according to their content type

### PR DESCRIPTION
**Pull request by claponcet**
# What Does This Do

Dispatches Lambda AppSec request bodies on their declared `Content-Type` instead of handing every body to a JSON parser:

| Declared type | Reported to the WAF as |
|---|---|
| JSON-ish (`application/json`, any `+json` suffix, `application/x-amz-json-1.1`, `application/javascript`) or no usable type at all | best-effort JSON parse |
| `application/x-www-form-urlencoded` | multimap, the same shape already used for query parameters |
| `multipart/*` | form fields, each part dispatched on its own `Content-Type`; file parts contribute their filename, not their content |
| `text/*` and everything else | raw string |

A body is never dropped. Any type we cannot structure, and any parse failure, degrades to the raw string, which the WAF can still match string rules against. In particular a `text/plain` body of `12345` must reach the WAF as a `String` and not as the `Double` a JSON parse would produce.

The same rule decides what happens to a malformed multipart body. A part whose headers run into the next delimiter is not reported part-by-part: no conforming parser accepts such a body — Commons FileUpload and Netty both reject it outright — so reporting the parts that happen to survive would show the WAF less than the application receives. The whole body degrades to the raw string instead.

Parsing is bounded by three allowances, shared across nesting levels rather than re-satisfied at each one: 1 MiB of characters read, 256 parts or parameters, and 20 levels of nesting. The depth and part limits mirror the WAF's own (`WAFModule.MAX_DEPTH` / `MAX_ELEMENTS`) — structure beyond them is discarded by `ObjectIntrospection` before the WAF sees it, so producing it would be wasted work. Exceeding an allowance degrades the body to a raw string rather than truncating it, since a truncated map would hide the dropped parameter from every rule.

**Behavior changes worth a reviewer's eye.** The JSON-ish decision is now made in one place, `ContentTypeBodyParser.isJsonOrUntyped`, which the response path in `LambdaEventParser` shares with the request path. It matches on the subtype alone, where the response path previously looked for `json` or `javascript` anywhere in the raw header. Three consequences:

- A `Content-Type` holding nothing but parameters — `; charset=utf-8` — counts as declaring no type, and so gets the same best-effort JSON parse an absent header gets. Pinned by a case in `dispatchesOnContentType`.
- A **response** body with a blank `content-type` now gets that best-effort JSON parse too. Previously only an absent header did, since `""` does not contain `json`.
- A **response** body whose type carries the word in a *parameter* — `multipart/form-data; boundary=--json` — no longer reaches the JSON parser. This is why the gate matches on the subtype only: a client-chosen boundary must not decide how the body is read.

# Motivation

Peer tracers already structure urlencoded and multipart Lambda bodies. Without it, a form-encoded or multipart request reaching a Lambda is visible to the WAF only as one opaque string, so any rule addressing `server.request.body` by key cannot match — the same request routed through a non-Lambda entrypoint would be inspected properly.

# Additional Notes

`MultipartSplitter` no longer collects a `HashMap` of every header per part. Only `Content-Disposition` and `Content-Type` are ever read, so the others are matched and dropped as the scan passes them: a body that is nothing but header lines would otherwise allocate a map entry per line, which is why the change is here at all. The naive version of the fix — matching header names inline with a case-insensitive `regionMatches` — measured slower on well-formed many-part forms, so the canonical spelling is tried with a case-sensitive compare first and the matching lives in its own small method rather than inlined into `split`.

Two known flaky failures were seen in `dd-trace-core` while validating, both unrelated to this diff and both passing on re-run in isolation: `PendingTraceBufferTest.testingTracerFlareDumpWithMultipleTraces` and `DDAgentWriterCombinedTest.statsdCommFailure`. Neither is annotated `@Flaky`.

File parts populate `server.request.body.filenames`, which comes almost free once the parts are split. The other two file addresses are left empty on purpose: the Lambda event delivers the body as one UTF-8-decoded string, so a file's bytes are already corrupted by the time we see them, and `files_content` would report that damage as if it were the upload. `files_field_names` waits with it.

The repeated callback-registration boilerplate in `LambdaAppSecHandler` is left for its own change.

# Contributor Checklist

- Format the title according to [the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any other useful labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- Avoid using `close`, `fix`, or [any linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- Update the [CODEOWNERS](https://github.com/DataDog/dd-trace-java/blob/master/.github/CODEOWNERS) file on source file addition, migration, or deletion
- Update [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) with any new configuration flags or behaviors
- Once approved, [use merge queue](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#merge-queue) to merge the PR

Jira ticket: [PROJ-IDENT]

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by claponcet** [comment]
@codex review

File: dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+   */
+  private static int nextDelimiter(final String body, final String delimiter, final int from) {
+    if (body.startsWith(delimiter, from)) {
+      return from;
+    }
+    for (int newline = body.indexOf('\n', from);
+        newline >= 0;
+        newline = body.indexOf('\n', newline + 1)) {
+      if (body.startsWith(delimiter, newline + 1)) {
+        return newline + 1;
```
[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java]: Fixed in 8521f8b.

`nextDelimiter` now requires the matched `--<boundary>` to actually end its line — RFC 2046 transport padding then a break, a `--` close delimiter, or the end of a truncated body. A line like `--x-not-a-boundary` is therefore data, and no longer terminates the part early. Verified that the payload following such a line now reaches `server.request.body`.

I applied the same guard to the delimiter check inside the header scan, which carried the same hiding vector through a different door: a bogus part header line such as `--xa: b` made `split` abandon the part and drop its entire content.

Regression is covered by `delimitsContentExactly` — it fails on the pre-fix parser.

**Comment by claponcet** [comment]
@codex review

File: dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+  /**
+   * A delimiter only delimits if its line ends there, RFC 2046 transport padding aside. A content
+   * line that merely starts with it — {@code --x-not-a-boundary} for a boundary of {@code x} — is
+   * data, and must not end the part early and hide the rest from the WAF.
+   *
+   * @param after the index just past the matched delimiter
+   */
+  private static boolean endsLine(final String body, final int after) {
+    return after == body.length() || body.startsWith("--", after) || lineStart(body, after) >= 0;
```
[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java]: Fixed in 693dfe0.

Confirmed: my earlier `endsLine` fix was incomplete rather than wrong. It stopped a *part* delimiter from matching on a prefix, but its close-delimiter clause accepted a trailing `--` as proof the line had ended, so `--x--not-a-close` still terminated the body and hid everything after it. `endsLine` now steps past those dashes and requires the line to actually end there, RFC 2046 transport padding aside:

```java
final int end = body.startsWith("--", after) ? after + 2 : after;
return end == body.length() || lineStart(body, end) >= 0;
```

A real `--x--\r\nepilogue`, and a truncated `--x--` at EOF, both still read as close delimiters. Covered by `delimitsContentExactly`, whose content now carries such a line and asserts the exact reported range; it fails on the pre-fix parser.

File: dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        }
+        if (newline < 0) {
+          cursor = length;
+          break;
+        }
+        cursor = newline + 1;
+      }
+      if (!headersComplete) {
+        // Body truncated inside the headers: there is no content to report
+        break;
```
[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java]: Fixed in 693dfe0. `split` now returns nothing here, so `ContentTypeBodyParser` keeps the raw body.

One correction to the framing: there is no content section behind the truncation, since the body ends inside the headers. What was being dropped is the headers themselves — and those are content in their own right, since a `filename` a rule could match lives there. So the loss is real, just narrower than "the truncated suffix".

It was also inconsistent in both directions at once: no mainstream parser accepts a body cut short mid-headers — Commons FileUpload throws and the request is rejected — so we were reporting earlier fields the application never processed while dropping that filename. Voiding the parse fixes the second and makes this uniform with the bail-out for a part whose headers run into the next delimiter. Covered by a new `splitsBodies` row, `--x\r\n\r\nv\r\n--x\r\nA: b`, which reported one part before the change.

File: dd-trace-core/src/main/java/datadog/trace/lambda/LambdaEventParser.java [hunk 4]
```diff
@@ -94,8 +96,8 @@ static LambdaRequestData parseEvent(String json) {
         case ALB_MULTI_VALUE:
           return extractAlbData(event, triggerType);
         default:
-          // Unsupported trigger: AppSec skips the invocation entirely, so there is nothing to
-          // extract. The trigger type is carried by the caller, not by this result.
+          // Unsupported trigger: returning EMPTY makes the caller skip the invocation, so there is
+          // nothing to extract sinc ethe event is not supported.
```
Comment (joeyzhao2018) [inline dd-trace-core/src/main/java/datadog/trace/lambda/LambdaEventParser.java]:
```suggestion
          // nothing to extract since the event is not supported.
```

File: dd-trace-core/src/main/java/datadog/trace/lambda/ContentTypeBodyParser.java [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+ * the raw {@link String}, which the WAF can still match string rules against.
+ */
+final class ContentTypeBodyParser {
+
+  private static final Logger log = LoggerFactory.getLogger(ContentTypeBodyParser.class);
+
+  // These bound the work done in this parser only: exceeding any of them degrades the body to a raw
+  // string rather than dropping content.
+  static final int MAX_BYTES = 1024 * 1024;
+  static final int MAX_PARTS = 256;
```
[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/ContentTypeBodyParser.java:33]: The mechanics here are correct — I checked the accounting in `ObjectIntrospection`, and the numbers match exactly:

```
MAX_ELEMENTS = 256, spent as root(1) + per entry key(1) + value(1) [+ list(1) for urlencoded]
urlencoded: 1 + 3n <= 256  ->   85 params
multipart:  1 + 2n <= 256  ->  127 scalar fields
```

Past that, later fields are dropped. So the finding is real, and the first counter-argument I had for it — "the raw fallback truncates too, at `MAX_STRING_SIZE`" — turns out not to hold, because that 4096 cap is **per string**. In the structured map every field value is its own short string, so nothing is string-truncated and the loss is purely the element cap. In the raw fallback the whole body is a single string, so the cap applies to the body as a whole. Those behave very differently, and the crossover is about field length rather than body size:

```
map covers:  the first 85 params, complete
raw covers:  the first 4096 characters ~= 4096/L params, where L = average param length
map wins when 85 > 4096/L   ->   L > ~48 chars
```

For the input described — many short one-value parameters — raw would indeed show the WAF more bytes than the map does.

Not changing it in this PR, for two reasons:

1. **It isn't specific to this parser.** Every body collector in the agent hands `ObjectIntrospection` a map subject to the same 256-element cap; the Netty collector and the Lambda extension don't work around it either. Making the Lambda path uniquely fall back to raw would diverge from all of them, and it would trade key-addressed matching (a rule scoped to a parameter name fires on the map, not on a flat string) for raw byte coverage — a product decision that shouldn't be made for one code path in isolation.
2. **Working around it here would mean duplicating WAF internals.** `MAX_ELEMENTS` and `MAX_STRING_SIZE` live in `dd-java-agent/appsec`, which `dd-trace-core` does not depend on. The parser would have to carry its own copies of both constants and re-implement the element accounting to predict the truncation — two sources of truth that will drift.

Worth noting the truncation is not silent: `ObjectIntrospection.convert` calls `setWafTruncated()` and reports `wafInputTruncated` telemetry, so these bodies are visible as truncated rather than quietly incomplete.

Also worth noting this PR is still a strict improvement on the previous behaviour for Lambda, where the body was *always* passed as a single raw string and therefore always subject to the 4096-character cap.

Tracking the general case as a follow-up covering all collectors rather than patching it here.

File: dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+   *     multipart
+   */
+  static List<Part> split(final String body, final String boundary, final int partBudget) {
+    final List<Part> parts = new ArrayList<>();
+    if (body == null || boundary == null || boundary.isEmpty() || partBudget <= 0) {
+      return parts;
+    }
+    final String delimiter = "--" + boundary;
+    final int length = body.length();
+    int position = body.startsWith(delimiter) ? 0 : nextDelimiter(body, delimiter, 0);
```
[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java]: Confirmed and fixed in 6297d2c426.

Reproduced it standalone first: with a body of `--x--not-a-close\r\n--x\r\n\r\nSECRET\r\n--x--`, the old code selected offset 0 without validating it, saw the `--` and stopped — 0 parts, so the whole body fell back to a raw string. The unconditional `nextDelimiter(body, delimiter, 0)` now validates the line ending at offset 0 like every other delimiter decision in the file, and the same body yields 1 part with `SECRET` addressable by field name.

While verifying this I audited all five delimiter-decision sites in `split`, and one more was wrong — not this one, but adjacent enough that it would have come back as a separate finding. A delimiter sitting at the *very start* of a part's content has no line break of its own, having spent the one that terminated the headers. RFC 2046's grammar (`body-part := MIME-part-headers [CRLF *OCTET]`) permits reading it either way; we and Netty read it as a delimiter, Commons FileUpload reads it as content. So `--x\r\nContent-Disposition: form-data; name=a\r\n\r\n--x--\r\n' OR 1=1 --\r\n--x--` reported `a=[]` while a FileUpload-based handler received the payload as the field's value. Same commit now reads it as content.

Both changes are measured rather than reasoned. I built a differential harness comparing this splitter against Netty's `HttpPostMultipartRequestDecoder` and Commons FileUpload over 260,604 generated bodies, with a one-directional invariant — reporting *more* bytes than the app's parser is acceptable, reporting fewer is a bypass:

```
                      raw fallbacks   invariant violations
before                   116,904            5,860
after                     82,424                6
```

The 6 residual cases are Netty-only readings that produce empty values, so no attacker bytes are hidden in any of them. All 11 realistic client bodies I checked (browser fields, `curl -F`, file upload, LF-only client, MIME preamble, JSON part, delimiter near-miss inside a value) parse identically before and after. The 1 MB adversarial-boundary case still completes in 5 ms, and a 2,003-body truncation/substitution sweep throws nothing.

Tests: `MultipartSplitterTest` covers the preamble case as `preamble looks like one`, and the content-start case with a pair of bodies that differ by exactly one CRLF — one reads as content, one reads as a delimiter. Both were mutation-verified to fail with the production change reverted.

[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java]: Confirmed and fixed in 6297d2c426.

Reproduced standalone: for `--x--not-a-close\r\n--x\r\n\r\nSECRET\r\n--x--` the old code took offset 0 without validating that the delimiter ended its line, read the `--` as a close delimiter and stopped at 0 parts, so the body degraded to a raw string and `SECRET` was no longer addressable by field name. RFC 2046 §5.1.1 requires the preamble to be ignored, which is what the app's parser does. The initial position now goes through the same validation as every other delimiter decision in the file.

The same audit turned up a second, adjacent case that this finding didn't cover: a delimiter at the very start of a part's content has no line break of its own, and Commons FileUpload reads it as content where we read it as a delimiter. That one is fixed in the same commit.

Both are measured against Netty's `HttpPostMultipartRequestDecoder` and Commons FileUpload over 260,604 generated bodies, with a one-directional invariant (reporting more bytes than the app's parser is fine; reporting fewer is a bypass):

```
                      raw fallbacks   invariant violations
before                   116,904            5,860
after                     82,424                6
```

The 6 remaining cases are Netty-only readings with empty values, so no attacker bytes are hidden. 11 realistic client bodies are unchanged, and the adversarial-boundary and mutation-sweep cases still pass.

File: dd-trace-core/src/main/java/datadog/trace/lambda/ContentTypeBodyParser.java [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+    final StringTokenizer tokenizer = new StringTokenizer(body, "&");
+    while (tokenizer.hasMoreTokens()) {
+      if (!context.takePart()) {
+        log.debug("Part allowance exhausted, keeping urlencoded body as a raw string");
+        return null;
+      }
+      final String pair = tokenizer.nextToken();
+      final int equals = pair.indexOf('=');
+      final String name = decode(equals == -1 ? pair : pair.substring(0, equals));
+      if (!name.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/ContentTypeBodyParser.java]: Good catch, fixed in 3c014692f8.

The old behaviour was pinned by a passing test (`skipsEmptyUrlEncodedPairsAndNames`), which asserted `&&=orphan&&a=1&&` produced a single-entry map — so `orphan` was dropped while the non-empty map suppressed the raw fallback, exactly as described. The `!name.isEmpty()` guard is gone; `""` is now a key like any other. Worth noting the `&&` case is unaffected either way: `StringTokenizer` skips empty tokens, so no pair is ever produced for it.

Your point about `NettyMultipartHelper` also applies one method over, so I fixed that in the same commit: `parseMultipart` had `if (name == null || name.isEmpty()) continue`, which dropped a part whose `Content-Disposition` carried `name=""` or a bare `name=`. Only `name == null` — no name parameter at all, so not a form field — is skipped now. That distinction is what `MultipartSplitter.parameter` already encodes: `null` for absent, `""` for present-but-empty.

Tests updated accordingly: `skipsEmptyUrlEncodedPairsButKeepsEmptyNames`, `keepsAMultipartFieldWithAnEmptyName`, and `treatsABareNameParameterAsAnEmptyName`. One existing test needed rewriting rather than extending — `skipsMultipartPartsWithoutAName` used `name=` as its example of a nameless part, which is now a field; it uses a part with no `name` parameter at all instead.

File: dd-trace-core/src/main/java/datadog/trace/lambda/LambdaEventParser.java (hunk 4 above)
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/LambdaEventParser.java]: Fixed in fad3363ad4, thanks.

**Comment by claponcet** [comment]
@codex review

File: dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+   * A delimiter only delimits if its line ends there, RFC 2046 transport padding aside — for the
+   * close delimiter, past its trailing {@code --}. A content line that merely starts with one, be
+   * it {@code --x-not-a-boundary} or {@code --x--not-a-close}, is data, and must not end the part
+   * early and hide the rest from the WAF.
+   *
+   * @param after the index just past the matched delimiter
+   */
+  private static boolean endsLine(final String body, final int after) {
+    final int end = body.startsWith("--", after) ? after + 2 : after;
+    return end == body.length() || lineStart(body, end) >= 0;
```
[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java]: Checked this against the two reference parsers on the exact body described (an ordinary delimiter with no trailing CRLF at EOF), and the current reading is the best of the three:

| body | ours | Netty `HttpPostMultipartRequestDecoder` | Commons FileUpload |
|---|---|---|---|
| ordinary delimiter at EOF | `{f1=[PAYLOAD]}` | `{f1=[PAYLOAD]}` | rejects the body |
| close delimiter at EOF | `{f1=[PAYLOAD]}` | `{f1=[PAYLOAD]}` | `{f1=[PAYLOAD]}` |
| ordinary delimiter + CRLF | raw fallback | `{f1=[PAYLOAD]}` | rejects the body |
| second part truncated | `{f1=…, f2=[SECOND]}` | `{f1=[PAYLOAD]}` | rejects the body |

`f1`'s value reaches the WAF intact in every case. The only bytes not reported are the delimiter text itself — which is derived from the `boundary` parameter of `Content-Type`, a header the WAF already inspects, and capped at 70 characters.

Both suggested remedies would push these bodies to the raw fallback, trading key-addressed matching on `f1` for the recovery of a few bytes of boundary text. That is a net loss, so leaving `endsLine` as it is.

File: dd-trace-core/src/main/java/datadog/trace/lambda/LambdaEventParser.java [hunk 9]
```diff
@@ -247,7 +240,8 @@ private static LambdaRequestData extractApiGatewayV1Data(Map<String, Object> eve
     if (queryParameters.isEmpty()) {
       queryParameters = extractQueryParameters(event.get("queryStringParameters"));
     }
-    Object body = extractBody(event);
+    ParseContext parseContext = new ParseContext();
+    Object body = extractBody(event, headers, parseContext);
```
[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/LambdaEventParser.java:237]: Fixed in b0e1025113.

One correction to the framing: the described trigger isn't a shape API Gateway sends. A REST proxy integration always populates both `headers` and `multiValueHeaders`, so `Content-Type` is never present only in the latter — the content-type dispatch was already reading it.

The real defect underneath is repeated-header collapse: `headers` keeps a single value per name, so for `Accept: text/html` + `Accept: application/json` the WAF was seeing one of the two. That is the same reason `multiValueQueryStringParameters` is already preferred over `queryStringParameters` a few lines below.

The merge is now a shared `mergeMultiValueHeaders` helper rather than a fourth inline copy — the ALB request path and the response path both use it too. That unified one behavioural difference between them: the response path used to drop a `multiValueHeaders` entry whose value wasn't a list, while the ALB path kept it. It now keeps it everywhere, since the runtime would still deliver that header and reporting less than the app sees is the failure mode that matters here.

File: dd-trace-core/src/main/java/datadog/trace/lambda/ContentTypeBodyParser.java [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+    }
+  }
+
+  /** Percent-decodes a single token, keeping it undecoded rather than dropping it on failure. */
+  private static String decode(final String value) {
+    if (value.isEmpty()) {
+      return value;
+    }
+    try {
+      return URLDecoder.decode(value, "UTF-8");
```
[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/ContentTypeBodyParser.java]: Fixed in cbeaf2e64e. `charsetName` now resolves the declared charset and `URLDecoder.decode` uses it, falling back to UTF-8 when none is declared, when the value is empty, or when it isn't a charset this JVM has.

Two things worth noting from the implementation:

- `MediaType.getCharset()` returns everything past `charset=`, so a following parameter and surrounding quotes come along with it (`iso-8859-1; q=1`, `"iso-8859-1"`). Normalising that is done locally rather than in `MediaType`, which is shared with other callers.
- `Charset.isSupported` throws `IllegalCharsetNameException` rather than returning `false` for a syntactically invalid name, so it needs a `try`/`catch` as well as the boolean.

Scope limit, to be honest about it: ASCII payloads percent-decode identically under both charsets, so this closes a non-ASCII gap rather than a general one. Ten tests in `ContentTypeBodyParserTest` cover it, including the unsupported and malformed declarations.

File: dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/MultipartSplitter.java:257]: The trace is accurate — `endsLine` does accept an ordinary delimiter at EOF, the preceding part is finalised before it, the next iteration finds no headers and merely advances, and `split` returns a non-empty list so the caller keeps the structured result. The bytes `\r\n--x` are reported nowhere. None of that is in dispute.

Declining it, because those bytes cannot conceal anything.

**Nothing can follow an EOF delimiter.** For a field to hide after it, the body would need header lines, a blank line and content — and then the delimiter would not be at EOF. The only unreported bytes are `--` plus the boundary token, which is read from the `boundary` parameter of `Content-Type` and capped at 70 characters by `extractBoundary`: a header the WAF already inspects. Every field before the truncation is reported with its value intact.

Both suggested remedies, measured on the exact body described:

```
--x\r\nContent-Disposition: form-data; name="f1"\r\n\r\nPAYLOAD\r\n--x

  current                                 {f1=[PAYLOAD]}
  accept EOF only for close delimiters    {f1=[PAYLOAD\r\n--x]}
  invalidate accumulated results          raw fallback (whole body as one string)
```

The first rejects nothing, despite the title: with the delimiter unrecognised, `next` becomes `-1`, `contentEnd` runs to the end of the body, and the bytes land inside `f1`'s value. It is neutral across the differential corpus I run this parser against — 260,604 generated bodies, unchanged at 82,424 raw fallbacks and 6 residual violations — but it gives `f1` a value that neither reference parser produces: Netty's `HttpPostMultipartRequestDecoder` reports `{f1=[PAYLOAD]}`, and Commons FileUpload rejects the body outright. The second trades key-addressed matching on every field in the body for the same few bytes of boundary text.

The asymmetry behind the finding is real, though, and worth making explicit: `split` detects a malformed structure in three places. Two of them `return Collections.emptyList()` — a part's headers running into the next delimiter, and the body ending inside a part's headers. This one advances `position` and continues. The distinction is what is at stake in each: those two find that part content or headers would go unreported, which is data an application would have parsed; this one finds only that a delimiter was truncated. Leaving `endsLine` as it is.

File: dd-trace-core/src/main/java/datadog/trace/lambda/LambdaEventParser.java [hunk 11]
```diff
@@ -534,6 +507,39 @@ private static Map<String, String> extractHeaders(Object headersObj) {
...
+    }
+    for (Map.Entry<?, ?> entry : ((Map<?, ?>) multiValueHeadersObj).entrySet()) {
+      if (entry.getKey() == null || entry.getValue() == null) {
+        continue;
+      }
+      // Lowercased for the same reason as extractHeaderMap
+      String key = String.valueOf(entry.getKey()).toLowerCase(Locale.ROOT);
+      if (entry.getValue() instanceof List) {
+        List<?> values = (List<?>) entry.getValue();
+        headers.put(key, values.stream().map(String::valueOf).collect(Collectors.joining(", ")));
```
Comment (joeyzhao2018) [inline dd-trace-core/src/main/java/datadog/trace/lambda/LambdaEventParser.java]:
```suggestion
        String separator = "cookie".equals(key) ? "; " : ", ";
        headers.put(key, values.stream().map(String::valueOf).collect(joining(separator)));
```
to take care of 'cookie' header since it uses ';'

File: dd-trace-core/src/main/java/datadog/trace/lambda/ContentTypeBodyParser.java [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+
+  /**
+   * Resolves the charset a form body declares, which decides what its percent-escapes mean: {@code
+   * %E9} is one character under ISO-8859-1 and an invalid sequence under UTF-8. Reading it the way
+   * the handler does keeps the WAF matching on the same value the application consumes.
+   *
+   * @return the declared charset, or UTF-8 when none is declared or it is not one this JVM has
+   */
+  private static String charsetName(final MediaType mediaType) {
+    String declared = mediaType.getCharset();
```
Comment (joeyzhao2018) [inline dd-trace-core/src/main/java/datadog/trace/lambda/ContentTypeBodyParser.java]: can we also use the `MultipartSplitter` here? like
```
String declared = MultipartSplitter.parameter(contentType, "charset");
```

Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/ContentTypeBodyParser.java]: Yes that simplifies things, thank you! Fixed

File: dd-trace-core/src/main/java/datadog/trace/lambda/LambdaEventParser.java (hunk 11 above)
Comment (claponcet) [reply dd-trace-core/src/main/java/datadog/trace/lambda/LambdaEventParser.java]: Fixed!

**Review by joeyzhao2018 — APPROVED** [review]
LGTM thanks
